### PR TITLE
[Merged by Bors] - feat(Data/Nat/Choose): Lucas' theorem

### PR DESCRIPTION
**Pull request by grhkm21**
This PR proves Lucas' theorem, which states that $\binom{n}{k} \equiv \prod_i \binom{\lfloor n / p^i \rfloor \pmod{p}}{\lfloor k / p^i \rfloor \pmod{p}} \equiv \prod_i \binom{n_i}{k_i} \pmod{p}$, where $n_i$ and $k_i$ are the $i$-th base-$p$ digits of $n$ and $k$ respectively. This is proven by looking at Freshman's dream closely + induction

### CODE REVIEW

File: Mathlib/Algebra/BigOperators/Basic.lean
Comment (b-mehta) [inline Mathlib/Algebra/BigOperators/Basic.lean:1118]: Hmm I'm not sure about this one, especially not as a simp lemma. I could just about believe that it's awkward to rewrite to get this by prod_ite_eq then h, but I worry that simp will try to prove `h` in cases where it's a big time sink

File: Mathlib/Algebra/BigOperators/Basic.lean
Comment (semorrison) [inline Mathlib/Algebra/BigOperators/Basic.lean:1329]: Write out if...then...else

Comment (semorrison) [reply Mathlib/Algebra/BigOperators/Basic.lean:1329]: and below

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (semorrison) [inline Mathlib/Data/Nat/Choose/Lucas.lean:30]: Couldn't this be an independent lemma?

Comment (semorrison) [reply Mathlib/Data/Nat/Choose/Lucas.lean:30]: I don't know the proof here, but the question above prompts asking if you can decompose this proof into further pieces!

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (semorrison) [inline Mathlib/Data/Nat/Choose/Lucas.lean:68]: Missing some mods in the statement here.

Comment (semorrison) [reply Mathlib/Data/Nat/Choose/Lucas.lean:68]: Should the statement mention the bounds hypotheses?

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (grhkm21) [reply Mathlib/Data/Nat/Choose/Lucas.lean:30]: This lemma is saying that $(X + 1)^n \equiv (X + 1)^{n \pmod{p}}(X^p + 1)^{\lfloor n / p \rfloor} \pmod{p}$ which follows because $(X + 1)^n \equiv (X + 1)^{n \pmod{p} + \lfloor n / p \rfloor \cdot p} \pmod{p}$ then use Freshman's dream. I will isolate it out into `Mathlib/Algebra/CharP/Basic.lean` then :D

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (grhkm21) [reply Mathlib/Data/Nat/Choose/Lucas.lean:68]: This is saying that $\binom{n}{k} \equiv \prod_i \binom{\lfloor n / p^i \rfloor}{\lfloor k / p^i \rfloor} \pmod{p}$ over all $i$ (but the floor eventually vanishes). Do you think the bounds requirement needs to be mentioned?

Comment (grhkm21) [reply Mathlib/Data/Nat/Choose/Lucas.lean:68]: Also I don't see where the "missing mods" are. It's saying that "..." is congruent to "product of "..." over i" modulo p.

Comment (semorrison) [reply Mathlib/Data/Nat/Choose/Lucas.lean:68]: The right hand side contains `choose (n / p ^ i % p) (k / p ^ i % p)`: both arguments have a `% p` which isn't reflected in your informal statement.

**Comment by semorrison** [comment]
(You can change the label back to `awaiting-review` when you're ready!)

**Comment by grhkm21** [comment]
Sorry for the long wait, exams :(

File: Mathlib/Algebra/BigOperators/Finsupp.lean
Comment (urkud) [inline Mathlib/Algebra/BigOperators/Finsupp.lean:110]: Is it a good `simp` lemma? Also, you rewrite many `ite`s as `if .. then .. else ..` but use `ite` here.

Comment (grhkm21) [reply Mathlib/Algebra/BigOperators/Finsupp.lean:110]: Yes, the `ite -> if ... then ... else ...` is my mistake. About the `simp` attribute, I just copied the definition above for `prod_ite_eq` without much thought. But now that I think about it, since the lemma requires a iff hypothesis `(h : ∀ x ∈ f.support, p x ↔ a = x)`, it won't ever be triggered by `simp` I think, or at least very very rarely. So I will remove the [simp].

File: Mathlib/Algebra/BigOperators/Group/Finset.lean
Comment (YaelDillies) [inline Mathlib/Algebra/BigOperators/Group/Finset.lean:2089]:
```suggestion
    (filter p s).card = ∑ a ∈ s, if p a then 1 else 0 := by simp [sum_ite]
```

File: Mathlib/Algebra/BigOperators/Group/Finset.lean
Comment (YaelDillies) [inline Mathlib/Algebra/BigOperators/Group/Finset.lean:2350]:
```suggestion
    (a : α) : ∏ i, (if p i then a else 1) = if ∃ i, p i then a else 1 := by
```

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (YaelDillies) [inline Mathlib/Data/Nat/Choose/Lucas.lean:27]: Please don't indent docstrings by default
```suggestion
choose (n / p) (k / p)` modulo `p`. -/
```

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (YaelDillies) [inline Mathlib/Data/Nat/Choose/Lucas.lean:55]: Why does this one not get a docstring?

File: Mathlib/Algebra/BigOperators/Group/Finset.lean
Comment (YaelDillies) [inline Mathlib/Algebra/BigOperators/Group/Finset.lean:1328]: I don't think we want that lemma? It's just a `prod_congr` away from an existing one

**Review by YaelDillies — COMMENTED** [review]
If you really care about the `if ... then ... else ...` changes, can you make them a separate PR? In fact, since I'm asking you to drop `prod_ite_eq_iff`, you can simply remove from this PR all the changes outside of `Data.Nat.Choose.Lucas` and `Algebra.CharP.Basic`

File: Mathlib/Algebra/BigOperators/Group/Finset.lean
Comment (grhkm21) [reply Mathlib/Algebra/BigOperators/Group/Finset.lean:1328]: IIRC I copied it from `Algebra/BigOperators/Finsupp.lean`. I will remove it

Comment (YaelDillies) [reply Mathlib/Algebra/BigOperators/Group/Finset.lean:1328]: Seems like you also added the `Algebra.BigOperators.Finsupp` one 🤔

Comment (grhkm21) [reply Mathlib/Algebra/BigOperators/Group/Finset.lean:1328]: lmfao sorry

**Comment by grhkm21** [comment]
I have reverted the `ite` -> `if-then-else` changes (TIL about `git reset --patch`). Hope it's okay now :D

**Review by YaelDillies — COMMENTED** [review]
You still have this lemma. Accident?

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (YaelDillies) [inline Mathlib/Data/Nat/Choose/Lucas.lean:64]:
```suggestion
theorem lucas_theorem' (a : ℕ) :
    choose n k ≡
      choose (n / p ^ a) (k / p ^ a) * ∏ i in range a, choose (n / p ^ i % p) (k / p ^ i % p) [ZMOD p] := by
```
so that it stays somewhat readable

Comment (grhkm21) [reply Mathlib/Data/Nat/Choose/Lucas.lean]: That makes the third line exceed 100 characters. I tried my best in the latest commit.

Comment (YaelDillies) [reply Mathlib/Data/Nat/Choose/Lucas.lean]: What about
```suggestion
theorem lucas_theorem' (a : ℕ) :
    choose n k ≡ choose (n / p ^ a) (k / p ^ a) *
      ∏ i in range a, choose (n / p ^ i % p) (k / p ^ i % p) [ZMOD p] := by
```
so that it stays somewhat readable

**Comment by grhkm21** [comment]
Okay I should have cleaned up stuff. Also I realised a proof didn't even compile, so I fixed that (it depended on the `prod_ite_eq_iff` thing

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (Parcly-Taxel) [inline Mathlib/Data/Nat/Choose/Lucas.lean:34]:
```suggestion
theorem choose_modEq_choose_mod_mul_choose_div : choose n k ≡ choose (n % p) (k % p) * choose (n / p) (k / p) [ZMOD p] := by
```

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (Parcly-Taxel) [inline Mathlib/Data/Nat/Choose/Lucas.lean:78]:
```suggestion
theorem choose_modEq_prod_range_choose {a : ℕ} (ha₁ : n < p ^ a) (ha₂ : k < p ^ a) :
```

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (Parcly-Taxel) [inline Mathlib/Data/Nat/Choose/Lucas.lean:64]:
```suggestion
theorem choose_modEq_choose_mul_prod_range_choose (a : ℕ) :
```

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (YaelDillies) [reply Mathlib/Data/Nat/Choose/Lucas.lean:34]: I prefer the semantic names over names obsessed with symbol-reading

Comment (grhkm21) [reply Mathlib/Data/Nat/Choose/Lucas.lean:34]: I find this quite unreadable personally, especially for a (decently) well-known theorem "in this area". I will make these changes but add an alias below.

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (Parcly-Taxel) [inline Mathlib/Data/Nat/Choose/Lucas.lean:19]:
```suggestion
* `choose_modEq_prod_range_choose`: the binomial coefficient `n choose k` is congruent to the
```

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (YaelDillies) [inline Mathlib/Data/Nat/Choose/Lucas.lean:34]: I have a hard time thinking this unambiguously refers to Lucas theorem

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (YaelDillies) [reply Mathlib/Data/Nat/Choose/Lucas.lean]: @Parcly-Taxel, please don't merge in your own suggestions on someone else's PR when they are potentially contentious

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (grhkm21) [reply Mathlib/Data/Nat/Choose/Lucas.lean:34]: It's (somewhat trivially) equivalent to the product version, and potentially useful at times, but I can remove the theorem label for it.

Comment (YaelDillies) [reply Mathlib/Data/Nat/Choose/Lucas.lean:33]: Oh I meant the lemma name, not the statement

**Review by Parcly-Taxel — APPROVED** [review]

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Nat/Choose/Lucas.lean:12]:
```suggestion
This file contains a proof of [Lucas's theorem](https://en.wikipedia.org/wiki/Lucas's_theorem) about
```

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Nat/Choose/Lucas.lean:82]:
```suggestion
  simp_rw [Nat.div_eq_of_lt ha₁, Nat.div_eq_of_lt ha₂, choose, cast_one, one_mul, cast_prod,
    Int.ModEq.refl]
```

**Comment by grhkm21** [comment]
Sorry, my browser automatically corrects wikipedia links to wikiwand.

**Review by YaelDillies — APPROVED** [review]
maintainer merge

**Comment by jcommelin** [comment]
I take it that the `TODO` in the PR commit message no longer applies? If so, please update it.

**Comment by grhkm21** [comment]
@jcommelin Done, sorry about that

File: Mathlib/Algebra/CharP/Basic.lean
Comment (grhkm21) [reply Mathlib/Algebra/CharP/Basic.lean:132]: Oops I missed this, I will do it yes

Comment (grhkm21) [reply Mathlib/Algebra/CharP/Basic.lean:132]: @YaelDillies I think `p` doesn't have to be explicit since `x` and `y` would infer `R` which should uniquely infer `p` - at least mathematically the second part is true?

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (b-mehta) [inline Mathlib/Data/Nat/Choose/Lucas.lean:34]: Both sides of this are nats, right? So maybe it'd be nice to have versions using MOD p as well?

**Review by b-mehta — APPROVED** [review]

File: Mathlib/Algebra/CharP/Basic.lean
Comment (YaelDillies) [reply Mathlib/Algebra/CharP/Basic.lean:132]: Mathematically, yes. In practice, not really 😢

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (semorrison) [reply Mathlib/Data/Nat/Choose/Lucas.lean:34]: I agree with Bhavik here we should have the Nat version.

@grhkm21, would you mind either adding this, or simply saying "sorry, can someone do it in a subsequent PR", and then you can merge?

**Comment by semorrison** [comment]
bors d+

**Comment by grhkm21** [comment]
Woah I can accept a PR?

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (grhkm21) [reply Mathlib/Data/Nat/Choose/Lucas.lean:34]: How should I name the theorems? I will add some `_nat` and `_int`, but alternatively I can call it `zmod` vs `mod`, but that seems unpopular in Mathlib.

**Comment by grhkm21** [comment]
@b-mehta do you mind reviewing the latest (two) commits? I added the nat versions as suggested

File: Mathlib/Data/Nat/Choose/Lucas.lean
Comment (YaelDillies) [reply Mathlib/Data/Nat/Choose/Lucas.lean:34]: `modEq` and `intModEq`, maybe?

**Comment by Ruben-VandeVelde** [comment]
Looks like you got hit by the new linter:
```
warning: ././././Mathlib/Data/Nat/Choose/Lucas.lean:99:0: unclosed sections or namespaces; expected: '

end Choose'
```

**Comment by jcommelin** [comment]
Thanks a lot for this contribution! Let's just merge this. Follow-up versions can be done in a follow-up PR.

bors merge

**Pull request closed without merging**
